File: app/src/lib/push.js
import { get, writable } from 'svelte/store'

import { unwrap } from './api.js'
import {
  forgetPushSubscription,
  getPushKey,
  registerPushSubscription,
} from './generated/sdk.gen'

/**
 * Enrolling this browser for notifications, and letting it go.
 *
 * Phase one of `PUSH_NOTIFICATIONS_PROPOSAL.md`: nothing here sends anything.
 * What it does is the half that has to be right before sending is worth
 * writing — a browser the server can address, and can stop addressing.
 *
 * Two constraints shape all of it. Permission may only be asked for **from a
 * user gesture**, so nothing here runs on load; and on iOS the Push API exists
 * only in a web app added to the Home Screen, so a Safari tab has no
 * `PushManager` at all and must be told why rather than shown a dead switch.
 */

/** What this browser can do, as far as we can tell without asking anybody. */
export const pushSupport = writable({
  /** Whether the browser exposes the Push API here at all. */
  available: false,
  /** Whether the *server* has VAPID keys. Both halves are needed. */
  configured: false,
  /** `granted`, `denied`, `default`, or `unsupported`. */
  permission: 'unsupported',
  /** Whether this browser is registered with the server right now. */
  subscribed: false,
})

/**
 * Whether the Push API is reachable in this browsing context.
 *
 * On iOS this is false in an ordinary Safari tab and true in the same app added
 * to the Home Screen — which is the single most confusing thing about web push,
 * and the reason the setting explains itself rather than simply disappearing.
 *
 * @returns {boolean}
 */
export function pushAvailable() {
  return (
    typeof window !== 'undefined' &&
    'serviceWorker' in navigator &&
    'PushManager' in window &&
    'Notification' in window
  )
}

/** Turn the server's base64url VAPID key into the bytes `subscribe` wants. */
function keyBytes(base64url) {
  const padded = (base64url + '='.repeat((4 - (base64url.length % 4)) % 4))
    .replace(/-/g, '+')
    .replace(/_/g, '/')
  const raw = atob(padded)
  return Uint8Array.from(raw, (character) => character.charCodeAt(0))
}

/** A short name for this device, so a list of them can be told apart. */
function deviceLabel() {
  const agent = navigator.userAgent
  if (/iPhone|iPad|iPod/.test(agent)) return 'iPhone or iPad'
  if (/Android/.test(agent)) return 'Android'
  if (/Macintosh/.test(agent)) return 'Mac'
  if (/Windows/.test(agent)) return 'Windows'
  return 'This device'
}

/**
 * Read what this browser and this server can do between them.
 *
 * Safe on load: it asks for no permission and shows no prompt. It reports the
 * *existing* permission, which the browser hands over without a gesture.
 *
 * @returns {Promise<object>} The value now in `pushSupport`.
 */
export async function readPushSupport() {
  const available = pushAvailable()
  const key = await unwrap(() => getPushKey()).catch(() => null)
  const registration = available
    ? await navigator.serviceWorker.getRegistration()
    : null
  const existing = await registration?.pushManager?.getSubscription?.()
  const state = {
    available,
    configured: Boolean(key?.configured),
    permission: available ? Notification.permission : 'unsupported',
    subscribed: Boolean(existing),
  }
  pushSupport.set(state)
  return state
}

/**
 * Ask for permission and register this browser. **Call from a click.**
 *
 * @returns {Promise<{ok: boolean, reason?: string}>} Why, when it did not work.
 */
export async function enablePush() {
  if (!pushAvailable()) {
    return { ok: false, reason: 'This browser cannot receive notifications here.' }
  }
  const key = await unwrap(() => getPushKey()).catch(() => null)
  if (!key?.configured) {
    return { ok: false, reason: 'This server is not set up to send notifications.' }
  }

  // From the gesture that called this. Asked on load, browsers either refuse
  // outright or count it against the site.
  const permission = await Notification.requestPermission()
  if (permission !== 'granted') {
    await readPushSupport()
    return {
      ok: false,
      reason:
        permission === 'denied'
          ? 'Notifications are blocked for this site in the browser’s own settings.'
          : 'Notifications were not allowed.',
    }
  }

  const registration = await navigator.serviceWorker.ready
  // Re-used when there is one: `subscribe` returns the existing subscription
  // for the same key, and a different key throws rather than replacing it.
  const subscription =
    (await registration.pushManager.getSubscription()) ??
    (await registration.pushManager.subscribe({
      // Non-negotiable, and the reason push cannot carry silent data sync:
      // every message has to show the person something.
      userVisibleOnly: true,
      applicationServerKey: keyBytes(key.public_key),
    }))

  await sendSubscription(subscription)
  await readPushSupport()
  return { ok: true }
}

/**
 * Tell the server about a subscription this browser holds.
 *
 * Also called on launch, which is what prunes devices that stop coming back:
 * a push service cannot be relied on to report one as gone — Apple has been
 * seen answering `201` for an endpoint it had already replaced.
 *
 * @param {PushSubscription} subscription
 * @returns {Promise<void>}
 */
export async function sendSubscription(subscription) {
  const json = subscription.toJSON()
  await unwrap(() =>
    registerPushSubscription({
      body: {
        endpoint: json.endpoint,
        p256dh: json.keys.p256dh,
        auth: json.keys.auth,
        label: deviceLabel(),
      },
    })
  )
}

/**
 * Stop notifications for this browser, here and on the server.
 *
 * Both halves, and the server first: unsubscribing locally and failing to say
 * so leaves an endpoint that is pushed to forever and answers nothing.
 *
 * @returns {Promise<void>}
 */
export async function disablePush() {
  const registration = await navigator.serviceWorker?.getRegistration?.()
  const subscription = await registration?.pushManager?.getSubscription?.()
  if (subscription) {
    await unwrap(() =>
      forgetPushSubscription({ body: { endpoint: subscription.endpoint } })
    ).catch(() => {
      // Unreachable. Better to drop it here anyway than to leave the person
      // with a switch that will not turn off.
    })
    await subscription.unsubscribe()
  }
  await readPushSupport()
}

/**
 * Give up this browser's subscription because the session is ending.
 *
 * Signing out means "stop telling me about pomodoros", and a subscription
 * outlives a token: nothing about clearing an access token reaches the push
 * service, so a browser left enrolled goes on being pushed to by an account
 * nobody is signed in to. The same applies to signing in as somebody else,
 * where the device would otherwise keep the previous account's enrolment and
 * receive their notifications.
 *
 * **Order matters and the failure path is deliberate.** The server is told
 * first, while there is still a token to tell it with; then the browser
 * unsubscribes, which is the half that actually stops delivery and is done
 * whether or not the server could be reached. A device that cannot reach the
 * server on the way out still stops receiving, and the row it left behind is
 * pruned the next time something is sent to it.
 *
 * @returns {Promise<void>} When the browser is no longer subscribed.
 */
export async function purgePush() {
  // Local checks only before committing to any network work: on a browser that
  // was never enrolled — which is most of them, and every test — this returns
  // without a request.
  if (!pushAvailable() || Notification.permission !== 'granted') return
  try {
    await disablePush()
  } catch {
    // Never block a sign-out. Whatever went wrong, leaving is the thing the
    // person asked for.
  }
}

/**
 * Re-register on launch, if this browser is already enrolled.
 *
 * Silent by design: it asks for nothing and shows nothing, and does nothing at
 * all unless permission was granted earlier. What it buys is the `updated_at`
 * that says this device still exists.
 *
 * @returns {Promise<void>}
 */
export async function refreshSubscription() {
  // Everything cheap and local first. This runs on every app start, and asking
  // the server whether push is configured — for a browser that has never been
  // granted permission and so has nothing to refresh — is a request that
  // cannot change the answer. It also put a round trip in front of the first
  // paint, which is a cost paid by every page load to serve almost none.
  if (!pushAvailable() || Notification.permission !== 'granted') return
  const registration = await navigator.serviceWorker.getRegistration()
  const subscription = await registration?.pushManager?.getSubscription?.()
  if (!subscription) return
  await sendSubscription(subscription).catch(() => {})
  await readPushSupport()
}

File: backend/routers/push.py
"""Registering and forgetting the browsers that may be sent a notification.

Phase one of `PUSH_NOTIFICATIONS_PROPOSAL.md`: the subscription lifecycle, and
nothing that sends. Deliberately shippable on its own — a device can be enrolled
and un-enrolled, and that half is verifiable without a push service existing
anywhere.

Web push is the one optional secret in this application. Absent VAPID keys, this
router answers honestly rather than failing: see `Settings.vapid_private_key`.
"""

from fastapi import APIRouter, HTTPException, Response, status
from sqlalchemy import select

from config import get_settings
from deps import CurrentUser, DbSession
from models import PushSubscription
from schemas import PushEndpoint, PushKey, PushSubscriptionIn, PushSubscriptionOut

router = APIRouter(prefix="/push", tags=["Push"])

UNCONFIGURED = "This server is not set up to send notifications"
"""Explanation returned when a deployment has no VAPID keys."""


@router.get(
    "/key",
    response_model=PushKey,
    operation_id="getPushKey",
    summary="Whether push works here, and the key to subscribe with",
    description=(
        "A browser needs the server's VAPID public key before it can subscribe. "
        "A deployment without one reports `configured: false` rather than an "
        "error, because push being off is a configuration and not a fault."
    ),
)
def get_push_key(user: CurrentUser) -> PushKey:
    """Report whether push is configured, and the public key when it is.

    Parameters
    ----------
    user : User
        The authenticated user. Required only so the key is not a public fact
        about the deployment; it grants nothing on its own.

    Returns
    -------
    PushKey
        The public key, or nulls when the deployment has no VAPID keys.
    """
    settings = get_settings()
    if not settings.push_configured:
        return PushKey(configured=False, public_key=None)
    return PushKey(configured=True, public_key=settings.vapid_public_key)


@router.get(
    "/subscriptions",
    response_model=list[PushSubscriptionOut],
    operation_id="listPushSubscriptions",
    summary="The devices this account has enrolled",
)
def list_subscriptions(user: CurrentUser, db: DbSession) -> list[PushSubscription]:
    """List the authenticated user's registered browsers.

    Parameters
    ----------
    user : User
        The authenticated user.
    db : sqlalchemy.orm.Session
        Active database session.

    Returns
    -------
    list of PushSubscription
        Never another account's, and never carrying an endpoint.
    """
    return list(
        db.execute(
            select(PushSubscription)
            .where(PushSubscription.user_id == user.id)
            .order_by(PushSubscription.created_at)
        ).scalars()
    )


@router.post(
    "/subscriptions",
    response_model=PushSubscriptionOut,
    status_code=status.HTTP_201_CREATED,
    operation_id="registerPushSubscription",
    summary="Register this browser for notifications",
    description=(
        "Safe to call on every launch, which is what the client does: an "
        "endpoint already stored is updated rather than duplicated. That "
        "re-registration is also what prunes devices that stop coming back, "
        "since a push service cannot be relied on to report one as gone."
    ),
)
def register(
    payload: PushSubscriptionIn, user: CurrentUser, db: DbSession
) -> PushSubscription:
    """Store or refresh one browser's subscription.

    Parameters
    ----------
    payload : PushSubscriptionIn
        The subscription as the browser reports it.
    user : User
        The authenticated user.
    db : sqlalchemy.orm.Session
        Active database session.

    Returns
    -------
    PushSubscription
        The stored row.

    Raises
    ------
    fastapi.HTTPException
        With status 503 when the deployment has no VAPID keys. Storing an
        address nothing can ever send to would be worse than refusing it.
    """
    if not get_settings().push_configured:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=UNCONFIGURED
        )

    # Matched on the endpoint alone, not on (user, endpoint): the endpoint
    # belongs to a *browser*, so signing in as somebody else on the same device
    # has to move the subscription rather than add a second one. Two rows would
    # mean one person's notifications arriving on another person's screen.
    stored = db.execute(
        select(PushSubscription).where(PushSubscription.endpoint == payload.endpoint)
    ).scalar_one_or_none()

    subscription = stored or PushSubscription(endpoint=payload.endpoint)
    subscription.user_id = user.id
    subscription.p256dh = payload.p256dh
    subscription.auth = payload.auth
    subscription.label = payload.label
    if stored is None:
        db.add(subscription)
    db.commit()
    db.refresh(subscription)
    return subscription


@router.delete(
    "/subscriptions",
    status_code=status.HTTP_204_NO_CONTENT,
    operation_id="forgetPushSubscription",
    summary="Stop sending notifications to this browser",
    description=(
        "Idempotent. A browser whose permission was revoked reports it on its "
        "next launch, which may be long after the row was already pruned."
    ),
)
def forget(payload: PushEndpoint, user: CurrentUser, db: DbSession) -> Response:
    """Remove one of the authenticated user's subscriptions.

    Parameters
    ----------
    payload : PushEndpoint
        The endpoint to forget.
    user : User
        The authenticated user.
    db : sqlalchemy.orm.Session
        Active database session.

    Returns
    -------
    fastapi.Response
        Empty, with status 204, whether or not there was anything to remove.
    """
    stored = db.execute(
        select(PushSubscription).where(
            PushSubscription.endpoint == payload.endpoint,
            PushSubscription.user_id == user.id,
        )
    ).scalar_one_or_none()
    if stored is not None:
        db.delete(stored)
        db.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)
